File: file_extraction/xml_file_extractor.py
import os
from zipfile import ZipFile
import traceback



def extract_all_files():
    directory = "datasets/I20230502/I20230502/DESIGN/"
    export_path = "file_extraction/full_data/"
    for filename in os.listdir(directory):
        try:
            f = os.path.join(directory, filename)
            with ZipFile(f, 'r') as zObject:
                zObject.extractall(path=export_path)
            zObject.close()
        except Exception as err:
            print(err)
            traceback.print_exc()

def move_all_xml_files():
    export_path = "file_extraction/all_xml/"
    directory = "file_extraction/full_data/"
    for filename in os.listdir(directory):
        try:
            f = os.path.join(directory, filename)
            xml_file = f+"/"+filename+".XML"
            os.replace(xml_file,export_path+filename+".XML")
        except Exception as err:
            print(err)
            traceback.print_exc()

move_all_xml_files()

File: file_extraction/xml_file_extractor.go
package xml_file_extractor

import (
	"archive/zip"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

func extractAllFiles() error {
	directory := "datasets/I20230502/I20230502/DESIGN/"
	exportPath := "file_extraction/full_data/"

	files, err := os.ReadDir(directory)
	if err != nil {
		return err
	}

	for _, file := range files {
		filePath := filepath.Join(directory, file.Name())
		zipReader, err := zip.OpenReader(filePath)
		if err != nil {
			return err
		}
		defer zipReader.Close()

		for _, zipFile := range zipReader.File {
			zipFileReader, err := zipFile.Open()
			if err != nil {
				return err
			}
			defer zipFileReader.Close()

			extractFilePath := filepath.Join(exportPath, zipFile.Name)

			if zipFile.FileInfo().IsDir() {
				os.MkdirAll(extractFilePath, os.ModePerm)
			} else {
				outputFile, err := os.Create(extractFilePath)
				if err != nil {
					return err
				}
				defer outputFile.Close()

				_, err = io.Copy(outputFile, zipFileReader)
				if err != nil {
					return err
				}
			}
		}
	}

	return nil
}

func moveAllXMLFiles() error {
	exportPath := "file_extraction/all_xml/"
	directory := "file_extraction/full_data/"

	files, err := os.ReadDir(directory)
	if err != nil {
		return err
	}

	for _, file := range files {
		xmlFile := filepath.Join(directory, file.Name(), file.Name()+".XML")
		destPath := filepath.Join(exportPath, file.Name()+".XML")

		if err := os.Rename(xmlFile, destPath); err != nil {
			return err
		}
	}

	return nil
}

func _() {
	if err := extractAllFiles(); err != nil {
		fmt.Println("Error extracting files:", err)
		return
	}

	if err := moveAllXMLFiles(); err != nil {
		fmt.Println("Error moving XML files:", err)
		return
	}

	fmt.Println("XML files moved successfully.")
}
